### ISSUE TITLE
[go_router] Support asynchronous redirects

### ISSUE THREAD

**Issue by johnpryan** [flutter/flutter#105808]
This is a proposal to change GoRouterRedirect to an asynchronous function. 

```dart
typedef GoRouterRedirect = Future<String?> Function(GoRouterState state);
```

This would allow a redirect to make asynchronous calls, for example to check if the user is signed in:

```dart
redirect: (state) async {
  final signedIn = await _appState.auth.isSignedIn();
  if (!signedIn) return '/signin';
  return null;
},
```

**Comment by ahmednfwela** [comment]
in my opinion ALL go router navigation methods should return a `Future<void>` instead of just `void`
this includes:
`go`, `goNamed`, `push`, `pushNamed` and `pop`
and `GoRouterRedirect`

**Comment by ahmednfwela** [comment]
this will be unavoidable, since changing `redirect` to a future will lead to a chain of `Future`s being introduced:  `_getLocRouteMatchesWithRedirects`,`_push`,`_go`

**Comment by ahmednfwela** [comment]
@chunhtai there are common use cases where we want to wait for navigation to end
e. g. 
1. displaying a dialog in the second screen
2. accessing a global key in the second screen

**Comment by chunhtai** [comment]
@ahmednfwela These tasks are not well supported with the current API. and I think a better API would be providing onEnter/onExit callback on the GoRoute object instead of waiting on context.go. What do you think?

**Comment by johnpryan** [comment]
> Only if we wait for the future, but I am not sure why we need to because it is already an async operation to change to a new screen. i.e. the screen will only change on the next frame of calling go.context. I think we can just keep them as is.

If navigation is asynchronous, why wouldn't you return a Future?

**Comment by ahmednfwela** [comment]
onEnter/onExit are great in generalized contexts, specifically authorization.
however in these 2 cases, the responsibility for handling these cases goes to the caller, not the second page

Also, the internal API uses `Future`s but just hides them from the user
e.g.:
https://github.com/flutter/packages/blob/main/packages/go_router/lib/src/router.dart#L188
https://github.com/flutter/packages/blob/main/packages/go_router/lib/src/router.dart#L217

**Comment by chunhtai** [comment]
> If navigation is asynchronous, why wouldn't you return a Future?

The current context may have already been disposed after the context.go future finishes if the new location does not contain the current page as part of its page stack. It may be error prompt if user is not careful.

> however in these 2 cases, the responsibility for handling these cases goes to the caller, not the second page

In this case the second page should have two different launch options, one for no dialog one for dialog

**Comment by ahmednfwela** [comment]
the second page is a route, not a dialog
a common case is:
Route A: `/students/new`
Route B: `/students`

user is currently in Route A > after he finishes adding a new student, he goes to `Route B` and we display a dialog "A new student is added successfully"

this is not easily done with the current API

**Comment by chunhtai** [comment]
@ahmednfwela

you can use `go('/students', extra: {'displayDialog': true})`.

Besides the current context maybe disposed. there is also an issue that `context.go('/students')` maybe redirected. It may be fine for your use case, but it may be a dangerous API.

**Comment by ahmednfwela** [comment]
if we pass `'displayDialog': true` to `/students`, then we are making it responsible for displaying the dialog, this is bad practice since we now have the students screen tightly coupled with the dialog, and it will be much harder to customize the dialog later

**Comment by chunhtai** [comment]
There is nothing stopping you from passing in function callback as extra.

`go('/students', extra: {'onLoaded': () { showDialog();}})`

I still think exposing the future is not a good API for this use case as per my comment above. but either way, makes context.go to expose the future should be considered as a new feature request that can be separated from this issue. In this issue, I would like to focus on async redirect.

Feel free to file a new issue about exposing the future if you would like to continue the discussion.

**Comment by johnpryan** [comment]
I filed https://github.com/flutter/flutter/issues/108764 so we can discuss separately.

**Comment by iapicca** [comment]
isn't this conflicting with https://github.com/flutter/flutter/issues/99121 ?

**Comment by chunhtai** [comment]
@iapicca Can you explain more?

**Comment by iapicca** [comment]
if you pass [BuildContext to redirect](https://github.com/flutter/flutter/issues/99121) and also make redirect `async`
seems to conflict with [use_build_context_synchronously](https://dart-lang.github.io/linter/lints/use_build_context_synchronously.html) see also [Synchronous BuildContexts | Decoding Flutter ](https://youtu.be/bzWaMpD1LHY)
and without the possibility of calling `isMounted` looks error prone to me

**Comment by chunhtai** [comment]
@iapicca that is a good point. There is no way to check whether the context is stale or not in the API. I may need to think a bit how to approach this

**Comment by chunhtai** [comment]
I think async with the context is fine, it only become a problem if the context is used after an async gap. If developer tries to do it, they will get lint warning. It definitely need some documentation encourage people to use context look up before any async operation

**Comment by cedvdb** [comment]
A common use case is to reset the state after redirect:

```dart
await go('there');
resetForm()
```

if the navigation are async, not setting their return type as Future / Task / any async representation will be an issue eventually.

**Comment by NazarenoCavazzon** [comment]
I made a PR for returning a value on pop, and I made all the parameters asynchronous except for the go method https://github.com/flutter/packages/pull/2416

**Comment by chunhtai** [comment]
@cedvdb @NazarenoCavazzon we should continue the discussion in https://github.com/flutter/flutter/issues/108764

**Comment by chunhtai** [comment]
fix merged into feature branch v5, please use https://github.com/flutter/flutter/issues/108044 to track v5 release progress
